### PR TITLE
Ascend attention backend(PA&MLA)

### PR DESCRIPTION
**Pull request by ping1jing2**
## Motivation

Currently only `torch_native` backend works for NPU, it has cycle by batch that works slow for bigger batch, in this MR new Ascend Attention backend for NPU device was implemented. It has big performance improvement for big batch size.

## Modifications

- PA
    - New attention backend - `AscendAttnBackend`: support paged attention with page size 128 only.
    
       New option for server argument `attention-backend` - "ascend"
    
    - `AscendTokenToKVPool` for store KVCache in paged format with npu operation `torch_npu._npu_reshape_and_cache`
    
    - `AscendPagedTokenToKVPoolAllocator` for allocate position in KVCache buffer, current paged allocator implemented on triton that does not work on NPU now
    
    - Added test for ascend attention `backend test/srt/test_ascend_attention_backend.py`
- MLA

    - Added `NPUMLAPagedTokenToKVPool` to support MLA optimization
    - Modified **deepseekv2** model to run it on other than "cuda" platforms
    - Unit test for DS V2 added

- bugfix
   `self.cuda_graph_mem_usage = 0` in model_runner added for fix bug with showing output result in run_bench_offline_throughput

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

## Accuracy and performance result
Used model for test: LLama-3.1-8b-Instruct
- PA
    ```shell
    python -m unittest test_ascend_attention_backend.TestAscendAttnBackend.test_gsm8k
    ```
    ```shell
    Accuracy: 0.755
    Invalid: 0.001
    Latency: 97.738 s
    Output throughput: 1217.759
    ```
    ```shell
    python -m unittest test_ascend_attention_backend.TestAscendAttnBackend.test_latency
    ```
    ```shell
    ====== Offline Throughput Benchmark Result =======
    Backend: engine
    Successful requests: 1
    Benchmark duration (s): 7.43
    Total input tokens: 38
    Total generated tokens: 235
    Last generation throughput (tok/s): 29.52
    Request throughput (req/s): 0.13
    Input token throughput (tok/s): 5.12
    Output token throughput (tok/s): 31.64
    Total token throughput (tok/s): 36.76
    ```
- MLA
    ```shell
    python3 -m unittest test_ascend_mla_backend.TestAscendMLABackend.test_gsm8k
    ```
    ```shell
    Accuracy: 0.648
    Invalid: 0.000
    Latency: 278.336 s
    Output throughput: 57.722 token/s
    ```

    ```shell
    python3 -m unittest test_ascend_mla_backend.TestAscendMLABackend.test_latency
    ```shell
    #Output tokens: 235
    #Input tokens: 256
    #Output tokens: 16
    
    ====== Offline Throughput Benchmark Result =======
    Backend:                                 engine
    Successful requests:                     1
    Benchmark duration (s):                  35.14
    Total input tokens:                      38
    Total generated tokens:                  235
    Last generation throughput (tok/s):      6.85
    Request throughput (req/s):              0.03
    Input token throughput (tok/s):          1.08
    Output token throughput (tok/s):         6.69
    Total token throughput (tok/s):          7.77
    ==================================================
    ```

## Pre-commit check
image

### CODE REVIEW

File: python/sglang/srt/mem_cache/allocator.py [hunk 1]
```diff
@@ -540,3 +540,164 @@ def clear(self):
...
+        seq_lens - 1 + page_size - 1
+    ) // page_size
+    end_new_pages = torch.cumsum(num_new_pages, 0)
+    start_new_pages = end_new_pages - num_new_pages
+    for i in range(len(seq_lens)):
+        if num_new_pages[i]:
+            out_indices[i] = free_pages[start_new_pages[i]] * page_size
+        else:
+            out_indices[i] = last_loc[i] + 1
+    return num_new_pages
```
[the remark this answers is not in the document]
Comment (ping1jing2) [reply python/sglang/srt/mem_cache/allocator.py:617]: We have another project named [triton-ascend](https://gitee.com/ascend/triton-ascend). Currently, we are designing a plan for SGLang and will give a solution as sonn as possible.

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    def gen_attention_mask(self, max_seq_len: int, dtype=torch.float16):
+        mask_flag = torch.tril(
+            torch.ones((max_seq_len, max_seq_len), dtype=torch.bool)
+        ).view(max_seq_len, max_seq_len)
+        mask_flag = ~mask_flag
+        if dtype == torch.float16:
+            mask_value = torch.finfo(torch.float32).min
+        else:
+            mask_value = 1
```
[the remark this answers is not in the document]
Comment (Makcum888e) [reply python/sglang/srt/layers/attention/ascend_backend.py:42]: It exist at least two different scenarios of using attention mask, first it masked value by some big number and second masked all values where mask is 1. So our mask should be like this

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            .to(dtype)
+            .to(self.device)
+        )
+        self.mask_len = max_seq_len
+
+    def __init__(self, model_runner: ModelRunner):
+        super().__init__()
+        self.forward_metadata = ForwardMetadata()
+        self.device = model_runner.device
+        self.gen_attention_mask(128, model_runner.dtype)
```
[the remark this answers is not in the document]
Comment (Makcum888e) [reply python/sglang/srt/layers/attention/ascend_backend.py:56]: it's a _npu_flash_attention_qlens restriction

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            forward_batch.req_to_token_pool.req_to_token[
+                forward_batch.req_pool_indices, : forward_batch.seq_lens.max()
+            ][:, :: self.page_size]
+            // self.page_size
+        )
+        if forward_batch.extend_seq_lens is not None:
+            self.forward_metadata.extend_seq_lens_cpu_int = (
+                forward_batch.extend_seq_lens.cpu().int()
+            )
+        self.forward_metadata.seq_lens_cpu_int = forward_batch.seq_lens_cpu.int()
```
[the remark this answers is not in the document]
Comment (Makcum888e) [reply python/sglang/srt/layers/attention/ascend_backend.py:76]: all attention operations in torch_npu need seq_lens on cpu

File: python/sglang/srt/mem_cache/allocator.py [hunk 5]
```diff
@@ -540,3 +540,164 @@ def clear(self):
...
+    def __init__(
+        self,
+        size: int,
+        page_size: int,
+        dtype: torch.dtype,
+        device: str,
+        kvcache: KVCache,
+    ):
+        super().__init__(size, page_size, dtype, device, kvcache)
+        self.ret_values = torch.empty((), dtype=torch.int32, device=self.device)
```
[the remark this answers is not in the document]
Comment (Makcum888e) [reply python/sglang/srt/mem_cache/allocator.py:631]: this code can not use Triton

**Comment by ping1jing2** [comment]
@ispobock  @Alcanderian  @zhyncs  Please help trigger the CI. Thank you.

File: test/srt/test_ascend_attention_backend.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                parallel=128,
+                host=f"http://{url.hostname}",
+                port=int(url.port),
+            )
+
+            metrics = run_eval_few_shot_gsm8k(args)
+            self.assertGreaterEqual(metrics["accuracy"], 0.62)
+            self.assertLessEqual(metrics["latency"], 150)
+        finally:
+            kill_process_tree(process.pid)
```
[the remark this answers is not in the document]
Comment (Makcum888e) [reply test/srt/test_ascend_attention_backend.py:70]: for CI this tests will be enough

File: python/sglang/srt/server_args.py [hunk 7]
```diff
@@ -378,6 +378,12 @@ def __post_init__(self):
             )
             self.disable_cuda_graph = True
 
+        if self.attention_backend == "ascend":
+            logger.warning(
+                "At this moment Ascend attention backend only supports a page_size of 128, change page_size to 128."
+            )
+            self.page_size = 128
```
[the remark this answers is not in the document]
Comment (VDV1985) [reply python/sglang/srt/server_args.py:387]: It is done the same way as for cutlass_mla backend. Some torch_npu attention operaitions may raise an error for page_size != 128

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 8]
```diff
@@ -568,6 +568,76 @@ def set_kv_buffer(
...
+            if v_scale is not None:
+                cache_v.div_(v_scale)
+            cache_k = cache_k.to(self.dtype)
+            cache_v = cache_v.to(self.dtype)
+
+        if self.store_dtype != self.dtype:
+            cache_k = cache_k.view(self.store_dtype)
+            cache_v = cache_v.view(self.store_dtype)
+
+        import torch_npu
```
[the remark this answers is not in the document]
Comment (ping1jing2) [reply python/sglang/srt/mem_cache/memory_pool.py:626]: this code is a platform specific and moving import will break all other platforms

File: python/sglang/srt/model_executor/model_runner.py [hunk 9]
```diff
@@ -302,6 +306,7 @@ def initialize(self, min_per_gpu_memory: float):
             self.init_cuda_graphs()
         else:
             self.cuda_graph_runner = None
+            self.cuda_graph_mem_usage = 0
```
[the remark this answers is not in the document]
Comment (ping1jing2) [reply python/sglang/srt/model_executor/model_runner.py:316]: `self.cuda_graph_mem_usage = 0` in model_runner added for fix bug with showing output result in run_bench_offline_throughput

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 10]
```diff
@@ -739,9 +746,16 @@ def compute_position_torch(
...
+if _is_npu:
+
+    def clamp_position(seq_lens):
+        return torch.clamp((seq_lens - 1), min=0).to(torch.int64)
+
+else:
+
+    @torch.compile(dynamic=True, backend=get_compiler_backend())
+    def clamp_position(seq_lens):
+        return torch.clamp((seq_lens - 1), min=0).to(torch.int64)
```
Comment (Alcanderian) [inline python/sglang/srt/model_executor/forward_batch_info.py]:
```suggestion
@torch.compile(dynamic=True, backend=get_compiler_backend(), disable=_is_npu)
def clamp_position(seq_lens):
    return torch.clamp((seq_lens - 1), min=0).to(torch.int64)
```

**Review by Alcanderian — APPROVED** [review]

File: python/sglang/srt/model_executor/forward_batch_info.py (hunk 10 above)
Comment (ping1jing2) [reply python/sglang/srt/model_executor/forward_batch_info.py]: it works and we have made the modifications

File: python/sglang/srt/models/deepseek_v2.py [hunk 11]
```diff
@@ -956,7 +956,9 @@ def _dispatch_mla_subtype():
                 else:
                     return AttnForwardMethod.MLA
 
-        if self.attention_backend == "flashinfer":
+        if self.attention_backend == "ascend":
+            return AttnForwardMethod.MLA
```
Comment (Alcanderian) [inline python/sglang/srt/models/deepseek_v2.py:960]: NOTE: There are some issues with MHA in npu backend so we only use MLA backend now
